src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/ResumableStateMachineStateAllocator.cs

```diff
@@ -0,0 +1,95 @@
+    internal sealed class ResumableStateMachineStateAllocator
+    {
+        private readonly VariableSlotAllocator? _slotAllocator;
+        private readonly bool _increasing;
+        private readonly int _firstState;
+
+        /// <summary>
+        /// The number of the next generated resumable state (i.e. state that resumes execution of the state machine after await expression or yield return).
+        /// </summary>
+        private int _nextState;
+
+#if DEBUG
+        /// <summary>
+        /// EnC support: states in this state machine that were matched to states of the previous generation state machine.
+        /// </summary>
+        private BitVector _matchedStates = BitVector.Empty;
+#endif
+        /// <summary>
+        /// EnC support: number of states in this state machine that match states of the previous generation state machine.
+        /// </summary>
+        private int _matchedStateCount;
+
+        public ResumableStateMachineStateAllocator(VariableSlotAllocator? slotAllocator, int firstState, bool increasing)
+        {
+            _increasing = increasing;
+            _slotAllocator = slotAllocator;
+            _matchedStateCount = 0;
+            _firstState = firstState;
+            _nextState = slotAllocator?.GetFirstUnusedStateMachineState(increasing) ?? firstState;
+        }
+
+        public int AllocateState(SyntaxNode awaitOrYieldReturnSyntax)
+        {
+            Debug.Assert(SyntaxBindingUtilities.BindsToResumableStateMachineState(awaitOrYieldReturnSyntax));
+
+            int direction = _increasing ? +1 : -1;
+
+            if (_slotAllocator?.TryGetPreviousStateMachineState(awaitOrYieldReturnSyntax, out var stateNumber) == true)
+            {
+#if DEBUG
+                // two states of the new state machine should not match the same state of the previous machine:
+                Debug.Assert(!_matchedStates[stateNumber * direction]);
+                _matchedStates[stateNumber * direction] = true;
+#endif
+                _matchedStateCount++;
+            }
+            else
+            {
+                stateNumber = _nextState;
+                _nextState += direction;
+            }
+
+            return stateNumber;
+        }
+
+        /// <summary>
+        /// True if any of the states generated for any previous state machine has not been allocated in this version.
+        /// </summary>
+        public bool HasMissingStates
+            => _matchedStateCount < (_slotAllocator?.GetFirstUnusedStateMachineState(_increasing) ?? _firstState) - _firstState;
+
+        public BoundStatement? GenerateThrowMissingStateDispatch(SyntheticBoundNodeFactory f, BoundExpression cachedState, string message)
+        {
+            if (!HasMissingStates)
+            {
+                return null;
+            }
+
+            return f.If(
+                f.Binary(
+                    _increasing ? BinaryOperatorKind.IntGreaterThanOrEqual : BinaryOperatorKind.IntLessThanOrEqual,
```

Comment:
>_increasing ? BinaryOperatorKind.IntGreaterThanOrEqual : BinaryOperatorKind.IntLessThanOrEqual [](http://example.com/codeflow?start=20&length=94)

Are we testing both of these code paths? (From the tests in EditAndContinueStateMachineTests, it looks like all the cases were for `_increasing: true` if I understood the code.)

Reply:
Updated IAsyncEnumerable test to cover this.